python/cudf/cudf/dataframe/column.py

```diff
@@ -61,6 +62,28 @@ def _concat(cls, objs, dtype=None):
             else:
                 return Column(Buffer.null(dtype))
 
+        # If all columns are `NumericalColumn` with different dtypes,
+        # we cast them to a common dtype.
+        # Notice, we can always cast pure null columns
+        not_null_cols = list(filter(lambda o: len(o) != o.null_count, objs))
+        if len(not_null_cols) > 0 and (
+            len(
+                [
+                    o
+                    for o in not_null_cols
+                    if not isinstance(o, NumericalColumn)
+                    or np.issubdtype(o.dtype, np.datetime64)
```

Comment:
Is this to prevent trying to concatenate say an `int16` with a `datetime64[s]`?

Reply:
Yes, in this case I think it makes sense to raise an exception